lib/main/atom/views/outline/navigationTreeComponent.tsx

```diff
@@ -0,0 +1,409 @@
+import {NavigationTreeViewModel, Props} from "./semanticViewModel"
+import {NavigationNodeComponent} from "./navigationNodeComponent"
+
+export class NavigationTreeComponent implements JSX.ElementClass {
+  private editor: TextEditor
+  public element: HTMLDivElement
+  private refs: {
+    main: HTMLOListElement
+  }
+  public selectedNode: NavigationTreeViewModel | null
+
+  private editorScrolling: Disposable
+  private editorChanging: Disposable
+  private activeEditorChanging: Disposable
+
+  constructor(public props: Props) {
+    this.selectedNode = null
+    this.prepareNavTree(props.navTree)
+    etch.initialize(this)
+    atom.workspace.observeActiveTextEditor(this.subscribeToEditor)
+  }
+
+  public async update(props: Partial<Props>) {
+    if (props.navTree) {
+      this.prepareNavTree(props.navTree)
+    }
+    this.props = {...this.props, ...props}
+    await etch.update(this)
+  }
+
+  public async destroy() {
+    if (this.editorScrolling) {
+      this.editorScrolling.dispose()
+    }
+    if (this.editorChanging) {
+      this.editorChanging.dispose()
+    }
+    if (this.activeEditorChanging) {
+      this.activeEditorChanging.dispose()
+    }
+    this.selectedNode = null
+    await etch.destroy(this)
+  }
+
+  private setEditor(editor: TextEditor) {
+    this.editor = editor
+  }
+
+  private async setNavTree(navTree: NavigationTreeViewModel | null) {
+    this.prepareNavTree(navTree)
+    if (isEqual(navTree, this.props.navTree)) {
+      return
+    }
+    this.restoreCollapsed(navTree, this.props.navTree)
+    this.props.navTree = navTree
+    this.selectedNode = null
+    await etch.update(this)
+  }
+
+  public async forceUpdate() {
+    await etch.update(this)
+  }
+
+  private async loadNavTree(filePath?: string) {
+    filePath = filePath ? filePath : this.editor.getPath()
+    if (filePath) {
+      try {
+        const client = await clientResolver.get(filePath)
+        await client.executeOpen({file: filePath})
+        const navtreeResult = await client.executeNavTree({file: filePath as string})
+        const navTree = navtreeResult ? navtreeResult.body! : undefined
+        if (navTree) {
+          this.setNavTree(navTree as NavigationTreeViewModel)
+        }
+      } catch (err) {
+        console.error(err, filePath)
+      }
+    }
+  }
+
+  /**
+   * HELPER modify / prepare NavigationTree for rendering.
+   *
+   * E.g. sort childItems by their location, preprocess className-string
+   *
+   * @param {NavigationTreeViewModel} navTree
+   *            the NavigationTree that will be prepared for rendering
+   */
+  private prepareNavTree(navTree: NavigationTreeViewModel | null): void {
+    if (navTree === null) return
+    navTree.styleClasses = this.getIconForKind(navTree.kind)
+    const modifiersClasses = this.getClassForKindModifiers(navTree.kindModifiers)
+    if (modifiersClasses) {
+      navTree.styleClasses += " " + modifiersClasses
+    }
```

Comment:
it's done here as a matter of "tree norminalization", before applying/rendering the new tree

but it can be moved into `NavigationNodeComponent` if necessary

Reply:
My point: why is assigning style classes to nodes is in "tree normalization" instead of in node presentation component?

Reply:
... OK, I am not sure anymore:  
I tried moving it into the constructor of `NavigationNodeComponent` (before calling `etch.initialize(this)`)

... but then I get rendering problems: at some point the specific styling for the nodes, i.e. the icons, do disappear

Any ideas where this should be done, so that there will be no rendering problems?

.... the first few seconds everything seem all right (but this is actually the deserialized tree):
![image](https://user-images.githubusercontent.com/3603062/35994729-9ed2dfa6-0d11-11e8-83b3-15080d5b20c5.png)

... but then it changes to this
![image](https://user-images.githubusercontent.com/3603062/35994873-0a2782de-0d12-11e8-80f7-ec9762211533.png)

looking at the HTML, I can see that the styling-classes vanish (or are not generated/set?), here the 
HTML code for one of the entries from the first screenshot:
```html
<li class="node entry exanded list-nested-item expanded selected">
  <div class="header list-item"><span class="icon icon-method modifier-public">refreshServiceUrl</span></div>
  <ol class="entries list-tree">
    <li class="node entry exanded list-item expanded">
      <div class="header list-item"><span class="icon icon-function">&lt;function&gt;</span></div>
      <ol class="entries list-tree"></ol>
    </li>
  </ol>
</li>
```

and here for the second:
```html
<li class="node entry exanded list-nested-item expanded selected">
  <div class="header list-item"><span class="undefined">refreshServiceUrl</span></div>
  <ol class="entries list-tree">
    <li class="node entry exanded list-item expanded">
      <div class="header list-item"><span class="undefined">&lt;function&gt;</span></div>
      <ol class="entries list-tree"></ol>
    </li>
  </ol>
</li>
```

Reply:
Try `NavigationNodeComponent.render`? I don't think this will have a performance impact. But if it would, then `NavigationNodeComponent.update` is as good a candidate as any.

If you're wondering, this is likely happening because Etch doesn't remove nodes if tags didn't change: https://github.com/atom/etch#keys

Reply:
ahh, of course, I also needed to update the style-classes in the `udpate` hook of the component -- OK, works now